<a href="https://colab.research.google.com/github/RyanHiatt/Project-Euler-Collection/blob/master/Hiatt_Project_Euler.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project Euler
My personal attempts at the Project Euler archives.

In [ ]:
import math
import numpy as np
import itertools
import pandas as pd
import operator
import functools
from collections import namedtuple
from fractions import Fraction
import sys

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive', force_remount=True)

Mounted at /content/gdrive


In [ ]:
!python --version

Python 3.7.15


## Problems 1 to 50

### Problem 1 - Multiples of 3 or 5
If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000.

In [ ]:
def sum_of_multiples(n1, n2, limit):
    return sum([i if i % n1 == 0 or i % n2 == 0 else 0 for i in range(limit)])


print(f"Answer 1 = {sum_of_multiples(n1=3, n2=5, limit=1000)}")

Answer 1 = 233168


### Problem 2 - Even Fibonacci Numbers
Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:

1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...

By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.

In [ ]:
# Create a fibonacci sequence from two starting values and a limit
def fibonacci_sequence(n1, n2, limit):
    fib_seq = [n1, n2]
    while True:
        if fib_seq[-1] + fib_seq[-2] <= limit:
            fib_seq.append(fib_seq[-1] + fib_seq[-2])
        else:
            return fib_seq


def even_fibonacci_sequence(n1, n2, limit):
    for element in fibonacci_sequence(n1=n1, n2=n2, limit=limit):
        if element % 2 == 0:
            yield element


print(f"Answer 2 = {sum(list(even_fibonacci_sequence(n1=1, n2=2, limit=4000000)))}")

Answer 2 = 4613732


### Problem 3 - Largest Prime Factor
The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?

In [ ]:
def get_prime_factors(number):
    while number % 2 == 0:
        yield 2
        number /= 2

    for i in range(3, int(math.sqrt(number) + 1), 2):
        while number % i == 0:
            yield i
            number /= i

    if number > 2:
        yield number


print(f"Answer 3 = {max(list(get_prime_factors(600851475143)))}")

Answer 3 = 6857


### Problem 4 - Largest Palindrome Product
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

In [ ]:
def largest_palindrome_product(start, stop):
    i_values = range(start, stop + 1)
    palindromes = []
    for i in i_values:
        j_values = (i for i in range(start, stop + 1))
        for j in j_values:
            if str(i * j) == str(i * j)[::-1]:
                palindromes.append(i * j)
    return max(palindromes)


print(f"Answer 4 = {largest_palindrome_product(100, 999)}")

Answer 4 = 906609


### Problem 5 - Smallest Multiple
2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is *evenly divisible* by all of the numbers from 1 to 20?

In [ ]:
def smallest_multiple(min_val, max_val):
    numbers = np.arange(start=min_val, stop=max_val + 1)
    multiple = np.array([max_val] * max_val)

    while True:
        if np.all(multiple % numbers == 0):
            return multiple[0]
        else:
            multiple += max_val


print(f"Answer 5 = {smallest_multiple(min_val=1, max_val=20)}")

Answer 5 = 232792560


### Problem 6 - Sum Square Difference
The sum of the squares of the first ten natural numbers is,  
$1^2+2^2+...+10^2=385$  
The square of the sum of the first ten natural numbers is,  
$(1+2+...+10)^2=55^2=385$  
Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is $3025-385=2640$.

Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

In [ ]:
def sum_square_difference(n):
    sum_of_square = 0
    for i in range(n + 1):
        sum_of_square += i * i

    square_of_sum = sum(range(n + 1)) ** 2

    return square_of_sum - sum_of_square


print(f"Answer 6 = {sum_square_difference(n=100)}")

Answer 6 = 25164150


### Problem 7 - 10,001st Prime
By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

What is the 10 001st prime number?

In [ ]:
# def sieve_of_eratosthenes(limit):  # Generate prime numbers up to a limit
#     a = [True] * limit  # Initialize the primality list
#     a[0] = a[1] = False
#
#     for (i, isprime) in enumerate(a):
#         if isprime:
#             yield i
#             for n in range(i*i, limit, i):  # Mark factors non-prime
#                 a[n] = False


def generate_primes(n_primes):  # Generate x prime numbers
    primes = []
    n = 2
    while len(primes) < n_primes:
        if all(n % i for i in itertools.islice(itertools.count(2), int(math.sqrt(n) - 1))):
            primes.append(n)
        n += 1
    return primes


print(f"Answer 7 = {generate_primes(n_primes=10001)[-1]}")

Answer 7 = 104743


### Problem 8 - Largest product in a series
The four adjacent digits in the 1000-digit number that have the greatest product are 9 × 9 × 8 × 9 = 5832.

73167176531330624919225119674426574742355349194934
96983520312774506326239578318016984801869478851843
85861560789112949495459501737958331952853208805511
12540698747158523863050715693290963295227443043557
66896648950445244523161731856403098711121722383113
62229893423380308135336276614282806444486645238749
30358907296290491560440772390713810515859307960866
70172427121883998797908792274921901699720888093776
65727333001053367881220235421809751254540594752243
52584907711670556013604839586446706324415722155397
53697817977846174064955149290862569321978468622482
83972241375657056057490261407972968652414535100474
82166370484403199890008895243450658541227588666881
16427171479924442928230863465674813919123162824586
17866458359124566529476545682848912883142607690042
24219022671055626321111109370544217506941658960408
07198403850962455444362981230987879927244284909188
84580156166097919133875499200524063689912560717606
05886116467109405077541002256983155200055935729725
71636269561882670428252483600823257530420752963450

Find the thirteen adjacent digits in the 1000-digit number that have the greatest product. What is the value of this product?

In [ ]:
number = """
7316717653133062491922511967442657474235534919493496983520312774506326239578318016984801869478851843
8586156078911294949545950173795833195285320880551112540698747158523863050715693290963295227443043557
6689664895044524452316173185640309871112172238311362229893423380308135336276614282806444486645238749
3035890729629049156044077239071381051585930796086670172427121883998797908792274921901699720888093776
6572733300105336788122023542180975125454059475224352584907711670556013604839586446706324415722155397
5369781797784617406495514929086256932197846862248283972241375657056057490261407972968652414535100474
8216637048440319989000889524345065854122758866688116427171479924442928230863465674813919123162824586
1786645835912456652947654568284891288314260769004224219022671055626321111109370544217506941658960408
0719840385096245544436298123098787992724428490918884580156166097919133875499200524063689912560717606
0588611646710940507754100225698315520005593572972571636269561882670428252483600823257530420752963450
""".replace("\n", "")


def largest_adj_product(series, num_digits):
    # Segment possible pools of numbers that do not contain zero
    num_pool = series.split("0")
    largest_product_series = None
    largest_product_obj = 0

    # check if the length of the segmented number is too short
    for index, n in enumerate(num_pool):
        if len(n) >= num_digits:
            # iterate through each group of num_digits within n
            for i in range(len(n) - num_digits + 1):
                iter_series = []
                iter_obj = 1

                # calculate iteration series and objective
                for j in range(i, i + num_digits):
                    iter_series.append(n[j])
                    iter_obj *= int(n[j])

                if iter_obj > largest_product_obj:  # Update series and objective
                    largest_product_series = iter_series
                    largest_product_obj = iter_obj

    return largest_product_series, largest_product_obj


s, p = largest_adj_product(series=number, num_digits=13)
print(f"Answer 8 = Series: {s}, Product: {p}")

Answer 8 = Series: ['5', '5', '7', '6', '6', '8', '9', '6', '6', '4', '8', '9', '5'], Product: 23514624000


### Problem 9 - Special Pythagorean triplet
A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,

$a^2 + b^2 = c^2$  
For example, $3^2 + 4^2 = 9 + 16 = 25 = 5^2$.

There exists exactly one Pythagorean triplet for which $a + b + c = 1000$.  
Find the product abc.

In [ ]:
# def primative_pythagorean_triplet_generator(hyp_max):
#     for m in range(2, hyp_max):
#         for n in range(m%2+1, m, 2):
#             a = m**2 - n**2
#             b = 2 * m * n
#             c = m**2 + n**2
#             if c > hyp_max:
#                 break
#             yield a, b, c

def pythagorean_triplet_generator(hyp_max):
    for m in range(2, hyp_max):
        for n in range(m%2+1, m, 2):
            for k in range(1, hyp_max):
                a = k * (m**2 - n**2)
                b = k * (2 * m * n)
                c = k * (m**2 + n**2)
                if c > hyp_max:
                    break
                yield a, b, c

def triplet_prod(target, hyp_max):
    for triplet in pythagorean_triplet_generator(hyp_max):
        if sum(triplet) == target:
            return f"Pythagorean Triplet: {triplet}, Sum = {sum(triplet)}, Product = {np.prod(triplet)}"

print(f"Answer 9 = {triplet_prod(1000, 500)}")

Answer 9 = Pythagorean Triplet: (375, 200, 425), Sum = 1000, Product = 31875000


### Problem 10 - Summation of primes
The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

Find the sum of all the primes below two million.

In [ ]:
def primes_sum(limit):  # Generate prime numbers up to a limit
    total = 0
    for n in range(2, limit+1):
        if all(n % i for i in itertools.islice(itertools.count(2), int(math.sqrt(n) - 1))):
            total += n
    return total


print(f"Answer 10 = {primes_sum(limit=2000000)}")

Answer 10 = 142913828922


### Problem 11 - Largest product in a grid
In the 20×20 grid below, four numbers along a diagonal line have been marked in red.

08 02 22 97 38 15 00 40 00 75 04 05 07 78 52 12 50 77 91 08  
49 49 99 40 17 81 18 57 60 87 17 40 98 43 69 48 04 56 62 00  
81 49 31 73 55 79 14 29 93 71 40 67 53 88 30 03 49 13 36 65  
52 70 95 23 04 60 11 42 69 24 68 56 01 32 56 71 37 02 36 91  
22 31 16 71 51 67 63 89 41 92 36 54 22 40 40 28 66 33 13 80  
24 47 32 60 99 03 45 02 44 75 33 53 78 36 84 20 35 17 12 50  
32 98 81 28 64 23 67 10 **26** 38 40 67 59 54 70 66 18 38 64 70  
67 26 20 68 02 62 12 20 95 **63** 94 39 63 08 40 91 66 49 94 21  
24 55 58 05 66 73 99 26 97 17 **78** 78 96 83 14 88 34 89 63 72  
21 36 23 09 75 00 76 44 20 45 35 **14** 00 61 33 97 34 31 33 95  
78 17 53 28 22 75 31 67 15 94 03 80 04 62 16 14 09 53 56 92  
16 39 05 42 96 35 31 47 55 58 88 24 00 17 54 24 36 29 85 57  
86 56 00 48 35 71 89 07 05 44 44 37 44 60 21 58 51 54 17 58  
19 80 81 68 05 94 47 69 28 73 92 13 86 52 17 77 04 89 55 40  
04 52 08 83 97 35 99 16 07 97 57 32 16 26 26 79 33 27 98 66  
88 36 68 87 57 62 20 72 03 46 33 67 46 55 12 32 63 93 53 69  
04 42 16 73 38 25 39 11 24 94 72 18 08 46 29 32 40 62 76 36  
20 69 36 41 72 30 23 88 34 62 99 69 82 67 59 85 74 04 36 16  
20 73 35 29 78 31 90 01 74 31 49 71 48 86 81 16 23 57 05 54  
01 70 54 71 83 51 54 69 16 92 33 48 61 43 52 01 89 19 67 48  

The product of these numbers is 26 × 63 × 78 × 14 = 1788696.

What is the greatest product of four adjacent numbers in the same direction (up, down, left, right, or diagonally) in the 20×20 grid?

In [ ]:

grid = """
08 02 22 97 38 15 00 40 00 75 04 05 07 78 52 12 50 77 91 08
49 49 99 40 17 81 18 57 60 87 17 40 98 43 69 48 04 56 62 00
81 49 31 73 55 79 14 29 93 71 40 67 53 88 30 03 49 13 36 65
52 70 95 23 04 60 11 42 69 24 68 56 01 32 56 71 37 02 36 91
22 31 16 71 51 67 63 89 41 92 36 54 22 40 40 28 66 33 13 80
24 47 32 60 99 03 45 02 44 75 33 53 78 36 84 20 35 17 12 50
32 98 81 28 64 23 67 10 26 38 40 67 59 54 70 66 18 38 64 70
67 26 20 68 02 62 12 20 95 63 94 39 63 08 40 91 66 49 94 21
24 55 58 05 66 73 99 26 97 17 78 78 96 83 14 88 34 89 63 72
21 36 23 09 75 00 76 44 20 45 35 14 00 61 33 97 34 31 33 95
78 17 53 28 22 75 31 67 15 94 03 80 04 62 16 14 09 53 56 92
16 39 05 42 96 35 31 47 55 58 88 24 00 17 54 24 36 29 85 57
86 56 00 48 35 71 89 07 05 44 44 37 44 60 21 58 51 54 17 58
19 80 81 68 05 94 47 69 28 73 92 13 86 52 17 77 04 89 55 40
04 52 08 83 97 35 99 16 07 97 57 32 16 26 26 79 33 27 98 66
88 36 68 87 57 62 20 72 03 46 33 67 46 55 12 32 63 93 53 69
04 42 16 73 38 25 39 11 24 94 72 18 08 46 29 32 40 62 76 36
20 69 36 41 72 30 23 88 34 62 99 69 82 67 59 85 74 04 36 16
20 73 35 29 78 31 90 01 74 31 49 71 48 86 81 16 23 57 05 54
01 70 54 71 83 51 54 69 16 92 33 48 61 43 52 01 89 19 67 48
""".replace("\n", " ").strip()

grid = map(int, grid.split(' '))
grid = np.array(list(grid)).reshape(20, 20)

def largest_product_in_grid(grid, num_adj):

    rows = len(grid)
    cols = len(grid[1])

    products = {}

    for i in range(rows):
        for j in range(cols):

            # Check right
            if j <= cols - num_adj:
                products[f'{i, j} Right'] = np.prod(grid[i, j:j+num_adj])
            
            # Check down
            if i <= rows - num_adj:
                products[f'{i, j} Down'] = np.prod(grid[i:i+num_adj, j])

            # Check down-right
            if i <= rows - num_adj and j <= cols - num_adj:
                arr = np.empty(num_adj, np.int_)
                for p in range(num_adj):
                    arr[p] = (grid[i+p, j+p])
                products[f'{i, j} Down-Right'] = np.prod(arr)

            # Check up-right
            if i >= num_adj - 1 and j <= cols - num_adj:
                arr = np.empty(num_adj, np.int_)
                for p in range(num_adj):
                    arr[p] = (grid[i-p, j+p])
                products[f'{i, j} Up-Right'] = np.prod(arr)

    max_prod_loc = max(products, key=products.get)
    max_prod = products[max_prod_loc]

    return max_prod, max_prod_loc

max_prod, location = largest_product_in_grid(grid, 4)
print(f"Answer 11 = {max_prod}, location: {location}")


Answer 11 = 70600674, location: (15, 3) Up-Right


### Problem 12 - Highly divisible triangular number
The sequence of triangle numbers is generated by adding the natural numbers. So the 7th triangle number would be 1 + 2 + 3 + 4 + 5 + 6 + 7 = 28. The first ten terms would be:

1, 3, 6, 10, 15, 21, 28, 36, 45, 55, ...

Let us list the factors of the first seven triangle numbers:

 1: 1  
 3: 1,3  
 6: 1,2,3,6  
10: 1,2,5,10  
15: 1,3,5,15  
21: 1,3,7,21  
28: 1,2,4,7,14,28  

We can see that 28 is the first triangle number to have over five divisors.

What is the value of the first triangle number to have over five hundred divisors?

In [ ]:
## Brut force method
# def get_factors(n):
#     for i in range(1, n + 1):
#         if n % i == 0:
#             yield i

# def triangle_number_generator():
#     return itertools.accumulate(itertools.count())

# def get_num_factors(number=None, target=None):
#     if number:
#         factors = get_factors(number)
#         num_factors = len(list(factors))
#         return num_factors

#     elif target:
#         tri_num = triangle_number_generator()
#         while True:
#             current_tri = next(tri_num)
#             num_factors = len(list(get_factors(current_tri)))
#             # print(current_tri, num_factors)
#             if num_factors >= target:
#                 return current_tri, num_factors

# triangular_number, num_factors = get_num_factors(target=500)
# print(f"Answer 12 = {triangular_number} with {num_factors} factors")


# Basic number theory
def triangle_number_generator():
    return itertools.accumulate(itertools.count())

def prime_factors(number):
    while number % 2 == 0:
        yield 2
        number /= 2

    for i in range(3, int(math.sqrt(number) + 1), 2):
        while number % i == 0:
            yield i
            number /= i

    if number > 2:
        yield int(number)

def get_num_factors(target):
    tri_num = triangle_number_generator()
    while True:
        current_tri = next(tri_num)
        if current_tri == 0:
            continue
        factors = prime_factors(current_tri)
        factor_count = 1
        for f, c in itertools.groupby(factors):
            factor_count *= (len(list(c)) + 1)
        if factor_count >= target:
            return current_tri, factor_count

tri_num, factor_count = get_num_factors(500)
print(f"Answer 12 = {tri_num} with {factor_count} factors")

Answer 12 = 76576500 with 576 factors


### Problem 13 - Large Sum
Work out the first ten digits of the sum of the following one-hundred 50-digit numbers.  

In [ ]:
number = """
37107287533902102798797998220837590246510135740250
46376937677490009712648124896970078050417018260538
74324986199524741059474233309513058123726617309629
91942213363574161572522430563301811072406154908250
23067588207539346171171980310421047513778063246676
89261670696623633820136378418383684178734361726757
28112879812849979408065481931592621691275889832738
44274228917432520321923589422876796487670272189318
47451445736001306439091167216856844588711603153276
70386486105843025439939619828917593665686757934951
62176457141856560629502157223196586755079324193331
64906352462741904929101432445813822663347944758178
92575867718337217661963751590579239728245598838407
58203565325359399008402633568948830189458628227828
80181199384826282014278194139940567587151170094390
35398664372827112653829987240784473053190104293586
86515506006295864861532075273371959191420517255829
71693888707715466499115593487603532921714970056938
54370070576826684624621495650076471787294438377604
53282654108756828443191190634694037855217779295145
36123272525000296071075082563815656710885258350721
45876576172410976447339110607218265236877223636045
17423706905851860660448207621209813287860733969412
81142660418086830619328460811191061556940512689692
51934325451728388641918047049293215058642563049483
62467221648435076201727918039944693004732956340691
15732444386908125794514089057706229429197107928209
55037687525678773091862540744969844508330393682126
18336384825330154686196124348767681297534375946515
80386287592878490201521685554828717201219257766954
78182833757993103614740356856449095527097864797581
16726320100436897842553539920931837441497806860984
48403098129077791799088218795327364475675590848030
87086987551392711854517078544161852424320693150332
59959406895756536782107074926966537676326235447210
69793950679652694742597709739166693763042633987085
41052684708299085211399427365734116182760315001271
65378607361501080857009149939512557028198746004375
35829035317434717326932123578154982629742552737307
94953759765105305946966067683156574377167401875275
88902802571733229619176668713819931811048770190271
25267680276078003013678680992525463401061632866526
36270218540497705585629946580636237993140746255962
24074486908231174977792365466257246923322810917141
91430288197103288597806669760892938638285025333403
34413065578016127815921815005561868836468420090470
23053081172816430487623791969842487255036638784583
11487696932154902810424020138335124462181441773470
63783299490636259666498587618221225225512486764533
67720186971698544312419572409913959008952310058822
95548255300263520781532296796249481641953868218774
76085327132285723110424803456124867697064507995236
37774242535411291684276865538926205024910326572967
23701913275725675285653248258265463092207058596522
29798860272258331913126375147341994889534765745501
18495701454879288984856827726077713721403798879715
38298203783031473527721580348144513491373226651381
34829543829199918180278916522431027392251122869539
40957953066405232632538044100059654939159879593635
29746152185502371307642255121183693803580388584903
41698116222072977186158236678424689157993532961922
62467957194401269043877107275048102390895523597457
23189706772547915061505504953922979530901129967519
86188088225875314529584099251203829009407770775672
11306739708304724483816533873502340845647058077308
82959174767140363198008187129011875491310547126581
97623331044818386269515456334926366572897563400500
42846280183517070527831839425882145521227251250327
55121603546981200581762165212827652751691296897789
32238195734329339946437501907836945765883352399886
75506164965184775180738168837861091527357929701337
62177842752192623401942399639168044983993173312731
32924185707147349566916674687634660915035914677504
99518671430235219628894890102423325116913619626622
73267460800591547471830798392868535206946944540724
76841822524674417161514036427982273348055556214818
97142617910342598647204516893989422179826088076852
87783646182799346313767754307809363333018982642090
10848802521674670883215120185883543223812876952786
71329612474782464538636993009049310363619763878039
62184073572399794223406235393808339651327408011116
66627891981488087797941876876144230030984490851411
60661826293682836764744779239180335110989069790714
85786944089552990653640447425576083659976645795096
66024396409905389607120198219976047599490197230297
64913982680032973156037120041377903785566085089252
16730939319872750275468906903707539413042652315011
94809377245048795150954100921645863754710598436791
78639167021187492431995700641917969777599028300699
15368713711936614952811305876380278410754449733078
40789923115535562561142322423255033685442488917353
44889911501440648020369068063960672322193204149535
41503128880339536053299340368006977710650566631954
81234880673210146739058568557934581403627822703280
82616570773948327592232845941706525094512325230608
22918802058777319719839450180888072429661980811197
77158542502016545090413245809786882778948721859617
72107838435069186155435662884062257473692284509516
20849603980134001723930671666823555245252804609722
53503534226472524250874054075591789781264330331690
""".strip().split("\n")

answer = int(str(sum([int(x) for x in number]))[:10])

print(f"Answer 13 = {answer}")

Answer 13 = 5537376230


### Problem 14 - Longest Collatz sequence
The following iterative sequence is defined for the set of positive integers:  

$n → n/2$ (n is even)  
$n → 3n + 1$ (n is odd)  

Using the rule above and starting with 13, we generate the following sequence:  

13 → 40 → 20 → 10 → 5 → 16 → 8 → 4 → 2 → 1  

It can be seen that this sequence (starting at 13 and finishing at 1) contains 10 terms. Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at 1.  

Which starting number, under one million, produces the longest chain?  

NOTE: Once the chain starts the terms are allowed to go above one million.  

In [ ]:
def collatz_sequence(n):
    yield n
    while n > 1:
        n = n // 2 if n % 2 == 0 else 3 * n + 1
        yield n

def modified_collatz(n):
    chain_length = 1
    while n > 1:
        n = n // 2 if n % 2 == 0 else 3 * n + 1
        chain_length += 1
    return chain_length

def longest_collatz_sequence(starting_value):
    counter = itertools.count(starting_value, -1)
    seq = []
    while True:
        i = next(counter)
        seq.append((i, modified_collatz(i)))
        if i == 1:
            break
    return max(seq, key=lambda x: x[1])

# print(list(collatz_sequence(12)))
print(f"Answer 14 = {longest_collatz_sequence(1000000)}")



Answer 14 = (837799, 525)


### Problem 15 - Lattice Paths
Starting in the top left corner of a 2×2 grid, and only being able to move to the right and down, there are exactly 6 routes to the bottom right corner.  

https://projecteuler.net/problem=15  

How many such routes are there through a 20×20 grid?  

In [ ]:
def binomial(a, b):
    return math.factorial(a) // math.factorial(b) // math.factorial(a - b)

print(f"Answer 15 = {binomial(40, 20)}")

Answer 15 = 137846528820


### Problem 16 - Power digit sum
$2^{15} = 32768$ and the sum of its digits is $3 + 2 + 7 + 6 + 8 = 26$.

What is the sum of the digits of the number $2^{1000}$?

In [ ]:
def sum_of_digits(n):
    return sum([int(x) for x in str(n)])

number = 2**1000
print(f"Answer 16 = {sum_of_digits(number)}")


Answer 16 = 1366


### Problem 17 - Number letter counts
If the numbers 1 to 5 are written out in words: one, two, three, four, five, then there are 3 + 3 + 5 + 4 + 4 = 19 letters used in total.  

If all the numbers from 1 to 1000 (one thousand) inclusive were written out in words, how many letters would be used?  


NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20 letters. The use of "and" when writing out numbers is in compliance with British usage.

In [ ]:
# one, two, three, four, five, six, seven, eight, nine, ten
# eleven, twelve, thirteen, fourteen, fifteen, sixteen, seventeen, eighteen, nineteen, twenty
# twenty, thirty, fourty, fifty, sixty, seventy, eighty, ninety
# one hundred ... nine hundred
# one thousand ... two thousand

# three hundred and forty two
# one hundred and fifteen
# nine hundred and twenty one


digits = {0: '', 1: 'one', 2: 'two', 3: 'three', 4: 'four', 5: 'five',
          6: 'six', 7: 'seven', 8: 'eight', 9: 'nine', 10: 'ten', 11: 'eleven', 
          12: 'twelve', 13: 'thirteen', 14: 'fourteen', 15: 'fifteen', 
          16: 'sixteen', 17: 'seventeen', 18: 'eighteen', 19: 'nineteen', 
          20: 'twenty', 30: 'thirty', 40: 'forty', 50: 'fifty', 60: 'sixty', 
          70: 'seventy', 80: 'eighty', 90: 'ninety'}


def num2word(num):
    num_str = str(num)
    l = len(num_str)
    word_str=''

    # 4th digit handling
    if l > 3:
        word_str += digits[int(num_str[-4])]
        word_str += 'thousand'

    # 3rd digit handling
    if l > 2:
        word_str += digits[int(num_str[-3])] if int(num_str[-3]) != 0 else ""
        word_str += 'hundred' if int(num_str[-3]) != 0 else ""

    # 'and' handling
    if l > 2 and (int(num_str[-2]) != 0 or int(num_str[-1]) != 0):
        word_str += 'and'

    # 2nd and 1st digit handling
    if l > 1:
        if int(num_str[-2]) == 1:
            word_str += digits[int(num_str[-2] + num_str[-1])] 
        elif int(num_str[-2]) == 2 and int(num_str[-1]) == 0:
            word_str += digits[20]
        elif int(num_str[-2]) > 2 and int(num_str[-1]) == 0:
            word_str += digits[int(num_str[-2] + num_str[-1])]
        else:
            word_str += digits[int(num_str[-2] + "0")]
            word_str += digits[int(num_str[-1])]

    # Single digit handling
    if l == 1:
        word_str += digits[int(num_str[-1])]

    return word_str


def letter_count(start, stop):
    letter_count = 0
    for i in range(start, stop + 1):
        letter_count += len(num2word(i))
    return letter_count

# for i in range(1, 1001):
#     print(num2word(i))

# print(num2word(9849))

print(f"Answer 17 = {letter_count(start=1, stop =1000)}")

Answer 17 = 21124


### Problem 18 - Maximum path sum I
By starting at the top of the triangle below and moving to adjacent numbers on the row below, the maximum total from top to bottom is 23.  

3  
7 4  
2 4 6  
8 5 9 3  

That is, $3 + 7 + 4 + 9 = 23$.  

Find the maximum total from top to bottom of the triangle below:

75  
95 64  
17 47 82  
18 35 87 10  
20 04 82 47 65  
19 01 23 75 03 34  
88 02 77 73 07 63 67  
99 65 04 28 06 16 70 92  
41 41 26 56 83 40 80 70 33  
41 48 72 33 47 32 37 16 94 29  
53 71 44 65 25 43 91 52 97 51 14  
70 11 33 28 77 73 17 78 39 68 17 57  
91 71 52 38 17 14 91 43 58 50 27 29 48  
63 66 04 68 89 53 67 30 73 16 69 87 40 31  
04 62 98 27 23 09 70 98 73 93 38 53 60 04 23  

NOTE: As there are only 16384 routes, it is possible to solve this problem by trying every route. However, Problem 67, is the same challenge with a triangle containing one-hundred rows; it cannot be solved by brute force, and requires a clever method! ;o)

In [ ]:
triangle = """
75
95 64
17 47 82
18 35 87 10
20 04 82 47 65
19 01 23 75 03 34
88 02 77 73 07 63 67
99 65 04 28 06 16 70 92
41 41 26 56 83 40 80 70 33
41 48 72 33 47 32 37 16 94 29
53 71 44 65 25 43 91 52 97 51 14
70 11 33 28 77 73 17 78 39 68 17 57
91 71 52 38 17 14 91 43 58 50 27 29 48
63 66 04 68 89 53 67 30 73 16 69 87 40 31
04 62 98 27 23 09 70 98 73 93 38 53 60 04 23
"""


def format_triangle(tri):
    # format into lists of integers
    tri = tri.strip().split('\n')
    for i, j in enumerate(tri):
        tri[i] = [int(x) for x in j.split()]
    
    # Find the number of rows and columns
    nrows = len(tri)
    ncols = max(len(x) for x in tri)

    # extend length of all arrays to be equal
    for row in tri:
        row.extend([0] * (ncols - len(row)))

    return np.array(tri), nrows, ncols


def max_path_sum(tri, nrows, ncols, print_tri=False):
    
    for i in range(nrows-2, -1, -1):
        if print_tri:
            print(tri, "\n")
        for j in range(ncols - 1):
            if tri[i,j] != 0:
                tri[i,j] += max(tri[i+1,j], tri[i+1,j+1])
        

    return tri[0][0]

            


tri, nrows, ncols = format_triangle(triangle) 
print(f"Answer 18 = {max_path_sum(tri, nrows, ncols)}")


Answer 18 = 1074


### Problem 19 - Counting Sundays
You are given the following information, but you may prefer to do some research for yourself.  

1 Jan 1900 was a Monday.  
Thirty days has September, April, June and November. All the rest have thirty-one, Saving February alone, Which has twenty-eight, rain or shine. And on leap years, twenty-nine.  
A leap year occurs on any year evenly divisible by 4, but not on a century unless it is divisible by 400.  

How many Sundays fell on the first of the month during the twentieth century (1 Jan 1901 to 31 Dec 2000)?

In [ ]:
month_codes = {1:0, 2:3, 3:3, 4:6, 5:1, 6:4, 7:6, 8:2, 9:5, 10:0, 11:3, 12:5}
century_codes = {17:4, 18:2, 19:0, 20:6, 21:4, 22:2, 23:0}
day_codes = {0: "Sunday", 1: "Monday", 2: "Tuesday", 3: "Wednesday", 4: "Thursday", 5: "Friday", 6: "Saturday"}

def day_of_week(day, month, year):
    yr = int(str(year)[-2:])

    year_code = (yr + yr // 4) % 7
    month_code = month_codes[month]
    century_code = century_codes[int(str(year)[:2])]
    leap_year = 0
    if month == 1 or month == 2:
        leap_year = 1 if year % 4 == 0 and year % 100 != 0 or year % 400 == 0 else 0

    day_code = (year_code + month_code + century_code + day - leap_year) % 7
    return day_codes[day_code]

def months_starting_on_day(start_year, end_year, day_str):
    cnt = 0
    for yr in range(start_year, end_year + 1):
        for mth in range(1, 13):
            if day_of_week(1, mth, yr) == day_str:
                cnt += 1
    return cnt


print(f"Answer 19 = {months_starting_on_day(1901, 2000, 'Sunday')}")

Answer 19 = 171


### Problem 20 - Factorial digit sum
$n!$ means $n × (n − 1) × ... × 3 × 2 × 1$  

For example, $10! = 10 × 9 × ... × 3 × 2 × 1 = 3628800$,
and the sum of the digits in the number $10!$ is $3 + 6 + 2 + 8 + 8 + 0 + 0 = 27$.  

Find the sum of the digits in the number 100!  

In [ ]:
def factorial_digit_sum(n):
    f = math.factorial(n)
    a = str(f)
    return sum([int(x) for x in a])

print(f"Answer 20 = {factorial_digit_sum(100)}")

Answer 20 = 648


### Problem 21 - Amicable numbers
Let $d(n)$ be defined as the sum of proper divisors of $n$ (numbers less than $n$ which divide evenly into $n$).  
If $d(a) = b$ and $d(b) = a$, where $a ≠ b$, then $a$ and $b$ are an amicable pair and each of $a$ and $b$ are called amicable numbers.  

For example, the proper divisors of 220 are 1, 2, 4, 5, 10, 11, 20, 22, 44, 55 and 110; therefore $d(220) = 284$. The proper divisors of 284 are 1, 2, 4, 71 and 142; so $d(284) = 220$.  

Evaluate the sum of all the amicable numbers under 10000.

In [ ]:
def sum_divisors(n):
    total = 1
    for x in range(2, int(math.sqrt(n) + 1)):
        if n % x == 0:
            total += x
            y = n // x
            if y > x:
                total += y
    return total


def find_amicable_number_sum(limit):
    amicable_nums = set()
    for a in range (2, limit):
        b = sum_divisors(a)
        c = sum_divisors(b)
        if a == c and a != b:
            if a not in amicable_nums:
                amicable_nums.add(a)
            if b not in amicable_nums:
                amicable_nums.add(b)

    return amicable_nums


# print(sum_divisors(28))
print(f"Answer 21 = {sum(find_amicable_number_sum(10000))}")


Answer 21 = 31626


### Problem 22 - Names scores
Using names.txt (right click and 'Save Link/Target As...'), a 46K text file containing over five-thousand first names, begin by sorting it into alphabetical order. Then working out the alphabetical value for each name, multiply this value by its alphabetical position in the list to obtain a name score.  

For example, when the list is sorted into alphabetical order, COLIN, which is worth 3 + 15 + 12 + 9 + 14 = 53, is the 938th name in the list. So, COLIN would obtain a score of 938 × 53 = 49714.  

What is the total of all the name scores in the file?

In [ ]:
with open("/content/drive/MyDrive/Colab_Projects/ProjectEuler/p022_names.txt") as f:
    names = sorted([x.lower() for x in f.read().replace("\"","").split(",")])
alphabet_value = {'a':1, 'b':2, 'c':3, 'd':4, 'e':5, 'f':6, 'g':7, 'h':8, 'i':9,
                  'j':10, 'k':11, 'l':12, 'm':13, 'n':14, 'o':15, 'p':16, 
                  'q':17,'r':18, 's':19, 't':20, 'u':21, 'v':22, 'w':23, 'x':24,
                  'y':25, 'z':26}

def name_scores(names):
    total = 0
    for i, name in enumerate(names):
        name_val = 0
        for letter in [*name]:
            name_val += alphabet_value[letter]
        # print(i+1, name, name_val)
        total += (i + 1) * name_val
    return total

print(f"Answer 22 = {name_scores(names)}")

Answer 22 = 871198282


### Problem 23 - Non-abundant sums
A perfect number is a number for which the sum of its proper divisors is exactly equal to the number. For example, the sum of the proper divisors of 28 would be 1 + 2 + 4 + 7 + 14 = 28, which means that 28 is a perfect number.

A number n is called deficient if the sum of its proper divisors is less than n and it is called abundant if this sum exceeds n.

As 12 is the smallest abundant number, 1 + 2 + 3 + 4 + 6 = 16, the smallest number that can be written as the sum of two abundant numbers is 24. By mathematical analysis, it can be shown that all integers greater than 28123 can be written as the sum of two abundant numbers. However, this upper limit cannot be reduced any further by analysis even though it is known that the greatest number that cannot be expressed as the sum of two abundant numbers is less than this limit.

Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers.

In [ ]:
def sum_divisors(n):
    total = 1
    for x in range(2, int(math.sqrt(n) + 1)):
        if n % x == 0:
            total += x
            y = n // x
            if y > x:
                total += y
    return total

def is_abundant(n):
    return True if n < sum_divisors(n) else False

def abundant_numbers(n):
    a = []
    for i in range(1, n):
        if is_abundant(i):
            a.append(i)
    return a

def non_abundant_sum(n):
    a = abundant_numbers(n)
    b = [0] * n
    for i in range(0, len(a)):
        for j in range(i, len(a)):
            sumofa = a[i] + a[j]
            if sumofa <= (n - 1):
                if b[sumofa] == 0:
                    b[sumofa] = sumofa

    total = 0
    for i in range(1, len(b)):
        if b[i] == 0:
            total += i

    return total


print(f"Answer 23 = {non_abundant_sum(28124)}")

Answer 23 = 4179871


### Problem 24 - Lexicographic permutations
A permutation is an ordered arrangement of objects. For example, 3124 is one possible permutation of the digits 1, 2, 3 and 4. If all of the permutations are listed numerically or alphabetically, we call it lexicographic order. The lexicographic permutations of 0, 1 and 2 are:  

012   021   102   120   201   210  

What is the millionth lexicographic permutation of the digits 0, 1, 2, 3, 4, 5, 6, 7, 8 and 9?

In [ ]:
millionth_permutation = int("".join([str(x) for x in list(itertools.permutations(range(10)))[999999]]))
print(f"Answer 24 = {millionth_permutation}")

Answer 24 = 2783915460


### Problem 25 - 1000-digit Fibonacci number
The Fibonacci sequence is defined by the recurrence relation:

$F_n = F_{n−1} + F_{n−2}$, where $F_1 = 1$ and $F_2 = 1$.
Hence the first 12 terms will be:

$F_1 = 1$  
$F_2 = 1$  
$F_3 = 2$  
$F_4 = 3$  
$F_5 = 5$  
$F_6 = 8$  
$F_7 = 13$  
$F_8 = 21$  
$F_9 = 34$  
$F_{10} = 55$  
$F_{11} = 89$  
$F_{12} = 144$  

The 12th term, F12, is the first term to contain three digits.

What is the index of the first term in the Fibonacci sequence to contain 1000 digits?

In [ ]:
def fibonacci_sequence_generator():
    fib_seq = [0, 1]
    yield 1
    while True:
        new = fib_seq[-1] + fib_seq[-2]
        fib_seq[-2] = fib_seq[-1]
        fib_seq[-1] = new
        yield new

def fib_index(target):
    fib = fibonacci_sequence_generator()
    ind = 0
    while True:
        current_fib = next(fib)
        ind += 1
        if len(str(current_fib)) >= target:
            return ind

print(f"Answer 25 = {fib_index(1000)}")

Answer 25 = 4782


### Problem 26 - Reciprocal cycles
A unit fraction contains 1 in the numerator. The decimal representation of the unit fractions with denominators 2 to 10 are given:  

$\frac{1}{2} = 0.5$  
$\frac{1}{3} = 0.(3)$  
$\frac{1}{4} = 0.25$  
$\frac{1}{5} = 0.2$  
$\frac{1}{6} = 0.1(6)$  
$\frac{1}{7} = 0.(142857)$  
$\frac{1}{8} = 0.125$  
$\frac{1}{9} = 0.(1)$  
$\frac{1}{10} = 0.1$  

Where 0.1(6) means 0.166666..., and has a 1-digit recurring cycle. It can be seen that $\frac{1}{7}$ has a 6-digit recurring cycle.

Find the value of d < 1000 for which $\frac{1}{d}$ contains the longest recurring cycle in its decimal fraction part.

In [ ]:
def recurring_cycle(n, d):
    # solve 10^s % d == 10^(s+t) % d
    # where t is length and s is start
    for t in range(1, d):
        if 1 == 10**t % d:
            return t
    return 0

longest = max(recurring_cycle(1, i) for i in range(2,1001))
print(f"Answer 26 = {[i for i in range(2,1001) if recurring_cycle(1, i) == longest][0]}")

Answer 26 = 983


### Problem 27 - Quadratic primes
Euler discovered the remarkable quadratic formula:  

$n^2+n+41$  

It turns out that the formula will produce 40 primes for the consecutive integer values $0\leq n\leq 39$. However, when $n=40, 40^2 + 40 + 41 = 40(40 + 1) + 41$ is divisible by 41, and certainly when $n=41, 41^2 + 41 + 41$ is clearly divisible by 41.  

The incredible formula $n^2 - 79n + 1601$ was discovered, which produces 80 primes for the consecutive values $0\leq n\leq 79$. The product of the coefficients, −79 and 1601, is −126479.  

Considering quadratics of the form:  

$n^2 + an + b$ where $|a| < 1000$ and $|b| \leq 1000$ where $|n|$ is the modulus/absolute value of $n$  

e.g. $|11| = 11$ and $|-4| = 4$  

Find the product of the coefficients, $a$ and $b$, for the quadratic expression that produces the maximum number of primes for consecutive values of $n$, starting with $n=0$.

In [ ]:
# n is the number to be check whether it is prime or not
n = 1

def is_prime(n):
    # this flag maintains status whether the n is prime or not
    prime_flag = 0
    if(n > 1):
        for i in range(2, int(math.sqrt(n)) + 1):
            if (n % i == 0):
                prime_flag = 1
                break
        if (prime_flag == 0):
            return True
        else:
            return False
    else:
        return False

def count_consecutive_primes(a, b):
    n = 0
    while True:
        if is_prime(n**2 + a * n + b):
            n += 1
        else:
            return n

def longest_quadratic_prime_coefficients():
    max_coef = (0, 0)
    max_primes = 0
    for a in range(-999, 1000):
        for b in range(-1000, 1001):
            c = count_consecutive_primes(a, b)
            if c > max_primes:
                max_coef = (a, b)
                max_primes = c
    return max_coef, max_primes
    
coeffs, prime_seq = longest_quadratic_prime_coefficients()
print(f"Answer 27 = {coeffs}, coefficients product: {coeffs[0] * coeffs[1]}, with longest prime sequence of {prime_seq}")

Answer 27 = (-61, 971), coefficients product: -59231, with longest prime sequence of 71


### Problem 28 - Number spiral diagonals
Starting with the number 1 and moving to the right in a clockwise direction a 5 by 5 spiral is formed as follows:  

21 22 23 24 25  
20  7  8  9 10  
19  6  1  2 11  
18  5  4  3 12  
17 16 15 14 13  

It can be verified that the sum of the numbers on the diagonals is 101.  

What is the sum of the numbers on the diagonals in a 1001 by 1001 spiral formed in the same way?

In [ ]:
def steps_from_center():
    Step  = namedtuple("Step", ["dx", "dy"])
    RIGHT = Step( 0,  1)
    DOWN  = Step( 1,  0)
    LEFT  = Step(0,  -1)
    UP    = Step( -1, 0)

    for n in itertools.count(start=1):
        if n % 2:
            yield RIGHT
            for i in range(n):
                yield DOWN
            for i in range(n):
                yield LEFT
        else:
            yield LEFT
            for i in range(n):
                yield UP
            for i in range(n):
                yield RIGHT

def create_number_spiral(rows, cols):
    # Initialize spiral grid
    spiral = np.array([[0]*cols]*rows)
    # Determine the maximum value
    max_i = rows * cols
    # Position
    x = rows // 2
    y = cols // 2
    # Set center to 1
    spiral[x, y] = 1

    # Create spiral
    for i, step in enumerate(steps_from_center(), start=2):
        if i > max_i:
            break
        else:
            x += step.dx
            y += step.dy
            spiral[x][y] = i

    return spiral

def calc_spiral_diagonals(rows, cols):
    spiral = create_number_spiral(rows, cols)
    reverse_spiral = np.fliplr(spiral)
    diag_sum = sum(np.diagonal(spiral)) + sum(np.diagonal(reverse_spiral)) - 1

    return spiral, diag_sum


spiral, diag_sum = calc_spiral_diagonals(1001, 1001)
# print(spiral)
print(f"Answer 28 = {diag_sum}")

[[1001001 1001002 1001003 ... 1001999 1002000 1002001]
 [1001000  997003  997004 ...  998000  998001  998002]
 [1000999  997002  993013 ...  994009  994010  998003]
 ...
 [1000003  996006  992017 ...  991021  995006  998999]
 [1000002  996005  996004 ...  995008  995007  999000]
 [1000001 1000000  999999 ...  999003  999002  999001]]
Answer 28 = 669171001


### Problem 29 - Distinct powers
Consider all integer combinations of $a^b$ for $2 ≤ a ≤ 5$ and $2 ≤ b ≤ 5$:  

$2^2=4, 2^3=8, 2^4=16, 2^5=32$  
$3^2=9, 3^3=27, 3^4=81, 3^5=243$  
$4^2=16, 4^3=64, 4^4=256, 4^5=1024$  
$5^2=25, 5^3=125, 5^4=625, 5^5=3125$  

If they are then placed in numerical order, with any repeats removed, we get the following sequence of 15 distinct terms:  

$4, 8, 9, 16, 25, 27, 32, 64, 81, 125, 243, 256, 625, 1024, 3125$  

How many distinct terms are in the sequence generated by $a^b$ for $2 ≤ a ≤ 100$ and $2 ≤ b ≤ 100$?

In [ ]:
def distinct_powers(x):
    unique_terms = set()

    for a in range(2, x+1):
        for b in range(2, x+1):
            unique_terms.add(a ** b)

    return len(unique_terms)


print(f"Answer 29 = {distinct_powers(100)}")

Answer 29 = 9183


### Problem 30 - Digit fifth powers
Surprisingly there are only three numbers that can be written as the sum of fourth powers of their digits:  

$1634 = 1^4 + 6^4 + 3^4 + 4^4$  
$8208 = 8^4 + 2^4 + 0^4 + 8^4$  
$9474 = 9^4 + 4^4 + 7^4 + 4^4$  
As $1 = 1^4$ is not a sum it is not included.  

The sum of these numbers is 1634 + 8208 + 9474 = 19316.  

Find the sum of all the numbers that can be written as the sum of fifth powers of their digits.

In [ ]:
def digit_power_sum(n, power):
    digits = [int(x) for x in str(n)]
    return sum(map(lambda x: x ** power, digits))

def solve(power):
    lbnd, ubnd = 10, 10 ** (power + 1)
    answer = 0
    for i in range(lbnd, ubnd + 1):
        if i == digit_power_sum(i, power):
            answer += i
    return answer

print(f"Anwer 30 = {solve(5)}")

Anwer 30 = 443839


### Problem 31 - Coin sums
In the United Kingdom the currency is made up of pound (£) and pence (p). There are eight coins in general circulation:  

1p, 2p, 5p, 10p, 20p, 50p, £1 (100p), and £2 (200p).  
It is possible to make £2 in the following way:  

1×£1 + 1×50p + 2×20p + 1×5p + 1×2p + 3×1p  
How many different ways can £2 be made using any number of coins?

In [ ]:
def num_coin_combinations(coins, target):
    combs = [0] * (target + 1)
    combs[0] = 1
    for c in coins:
        for i in range(c, target + 1):
            combs[i] += combs[i - c]
    return combs[target]

coins = [1, 2, 5, 10, 20, 50, 100, 200]
target = 200

print(f"Answer 31 = {num_coin_combinations(coins, target)}")

Answer 31 = 73682


### Problem 32 - Pandigital products
We shall say that an n-digit number is pandigital if it makes use of all the digits 1 to n exactly once; for example, the 5-digit number, 15234, is 1 through 5 pandigital.  

The product 7254 is unusual, as the identity, 39 × 186 = 7254, containing multiplicand, multiplier, and product is 1 through 9 pandigital.  

Find the sum of all products whose multiplicand/multiplier/product identity can be written as a 1 through 9 pandigital.  

HINT: Some products can be obtained in more than one way so be sure to only include it once in your sum.

In [ ]:
def is_pandigital(n, n_digits):
    n = set(list(str(n)))
    if len(n) == n_digits:
        return True
    else:
        return False

def pandigital_products(n_digits):
    prods = []
    for i in range(1, 100):
        for j in range(100, 10000):
            prod = str(i) + str(j) + str(i * j)
            if "0" not in prod and len(prod) == n_digits:
                if is_pandigital(prod, n_digits):
                    prods.append(i * j)
    return sum(set(prods))

print(f"Answer 32 = {pandigital_products(9)}")

Answer 32 = 45228


### Problem 33 - Digit cancelling fractions
The fraction 49/98 is a curious fraction, as an inexperienced mathematician in attempting to simplify it may incorrectly believe that 49/98 = 4/8, which is correct, is obtained by cancelling the 9s.  

We shall consider fractions like, 30/50 = 3/5, to be trivial examples.  

There are exactly four non-trivial examples of this type of fraction, less than one in value, and containing two digits in the numerator and denominator.  

If the product of these four fractions is given in its lowest common terms, find the value of the denominator.

In [ ]:
def digit_cancelling_fractions():
    dp = 1
    np = 1

    for c in range(1, 10):
        for d in range(1, c):
            for n in range(1, d):
                if (n * 10 + c) * d == (c * 10 + d) * n:
                    np *= n
                    dp *= d
    return dp / math.gcd(np, dp)

print(f"Answer 33 = {digit_cancelling_fractions()}")

Answer 33 = 100.0


### Problem 34 - Digit Factorials
145 is a curious number, as 1! + 4! + 5! = 1 + 24 + 120 = 145.  

Find the sum of all numbers which are equal to the sum of the factorial of their digits.  

Note: As 1! = 1 and 2! = 2 are not sums they are not included.

In [ ]:
def digit_factorial_sum(n):
    d = [int(x) for x in list(str(n))]
    fac_sum = sum(map(lambda x: math.factorial(x), d))
    return fac_sum

def solve():
    answer = 0
    for i in range(3, 1000000):
        s = digit_factorial_sum(i)
        if s == i:
            answer += i
    return answer

print(f"Answer 34 = {solve()}")

Answer 34 = 40730


### Problem 35 - Circular primes
The number, 197, is called a circular prime because all rotations of the digits: 197, 971, and 719, are themselves prime.  

There are thirteen such primes below 100: 2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, and 97.  

How many circular primes are there below one million?

In [ ]:
def is_prime(n):
    # this flag maintains status whether the n is prime or not
    prime_flag = 0
    if(n > 1):
        for i in range(2, int(math.sqrt(n)) + 1):
            if (n % i == 0):
                prime_flag = 1
                break
        if (prime_flag == 0):
            return True
        else:
            return False
    else:
        return False

def number_rotations(n) -> list:
    rotations = [n]
    n = [int(x) for x in list(str(n))]
    for i in range(len(n)-1):
        n = n[1:] + n[:1]
        rotations.append(int("".join([str(x) for x in n])))
    
    return rotations

def circular_primes(n_max):
    c_prime_count = 0
    for i in range(1, n_max):
        rotations = number_rotations(i)
        if all(map(lambda x: is_prime(x), rotations)):
            c_prime_count += 1
    return c_prime_count


print(f"Answer 35 = {circular_primes(1000000)}")

Answer 35 = 55


### Problem 36 - Double-base palindromes
The decimal number, 585 = 10010010012 (binary), is palindromic in both bases.  

Find the sum of all numbers, less than one million, which are palindromic in base 10 and base 2.  

(Please note that the palindromic number, in either base, may not include leading zeros.)

In [ ]:
def double_base_palindromes(n_max):
    n_sum = 0
    for i in range(1, n_max):
        if str(i) == str(i)[::-1]:
            if str(bin(i)[2:]) == str(bin(i)[:1:-1]):
                n_sum += i
    return n_sum

print(f"Answer 36 = {double_base_palindromes(1000000)}")

Answer 36 = 872187


### Problem 37 - Truncatable primes
The number 3797 has an interesting property. Being prime itself, it is possible to continuously remove digits from left to right, and remain prime at each stage: 3797, 797, 97, and 7. Similarly we can work from right to left: 3797, 379, 37, and 3.  

Find the sum of the only eleven primes that are both truncatable from left to right and right to left.  

NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes.

In [ ]:
def is_prime(n):
    # this flag maintains status whether the n is prime or not
    prime_flag = 0
    if(n > 1):
        for i in range(2, int(math.sqrt(n)) + 1):
            if (n % i == 0):
                prime_flag = 1
                break
        if (prime_flag == 0):
            return True
        else:
            return False
    else:
        return False


def left_truncate(n):
    s = [n]
    n = list(str(n))
    while len(n) > 1:
        n.pop(0)
        s.append(int("".join([str(x) for x in n])))
    return s


def right_truncate(n):
    s = [n]
    n = list(str(n))
    while len(n) > 1:
        n.pop()
        s.append(int("".join([str(x) for x in n])))
    return s


def two_sided_primes():
    count = 0
    answer = 0
    n = 10
    while count < 11:
        if all(map(lambda x: is_prime(x), left_truncate(n))):
            if all(map(lambda x: is_prime(x), right_truncate(n))):
                answer += n
                count += 1    
        n += 1
    return answer


print(f"Answer 37 = {two_sided_primes()}")

Answer 37 = 748317


### Problem 38 - Pandigital multiples
Take the number 192 and multiply it by each of 1, 2, and 3:  

192 × 1 = 192  
192 × 2 = 384  
192 × 3 = 576  

By concatenating each product we get the 1 to 9 pandigital, 192384576. We will call 192384576 the concatenated product of 192 and (1,2,3)  

The same can be achieved by starting with 9 and multiplying by 1, 2, 3, 4, and 5, giving the pandigital, 918273645, which is the concatenated product of 9 and (1,2,3,4,5).  

What is the largest 1 to 9 pandigital 9-digit number that can be formed as the concatenated product of an integer with (1,2, ... , n) where n > 1?

In [ ]:
def is_pandigital(n, n_digits):
    n = set(list(str(n)))
    if len(n) == n_digits:
        return True
    else:
        return False

def pandigital_multiple(n_digits):
    prod = 0
    for i in range(9487, 9234, -1):
        prod = str(i) + str(i * 2)
        if "0" not in prod and len(prod) == n_digits:
            if is_pandigital(prod, n_digits):
                return int(prod)

print(f"Answer 38 = {pandigital_multiple(9)}")

Answer 38 = 932718654


### Problem 39 - Integer right triangles
If p is the perimeter of a right angle triangle with integral length sides, {a,b,c}, there are exactly three solutions for p = 120.  

{20,48,52}, {24,45,51}, {30,40,50}  

For which value of p ≤ 1000, is the number of solutions maximised?

In [ ]:
def num_perimeter_solutions(p):
    count = 0
    for a in range(1, int(p/2)):
        for b in range(a, int(p/2)):
            c = p - a - b
            if a + b > c and a + c > b and b + c > a:
                if (a ** 2 + b ** 2) == c ** 2:
                    count += 1
    return count

def most_preimeter_solutions(max_p):
    perimeter = 0
    max_solutions = 0
    
    for i in range(1, max_p + 1):
        a = num_perimeter_solutions(i)
        if a > max_solutions:
            perimeter = i
            max_solutions = a
    
    return perimeter, max_solutions


p, sol = most_preimeter_solutions(1000)
print(f"Answer 39 = {p} with {sol} solutions")

Answer 39 = 840 with 8 solutions


### Problem 40 - Champernowne's constant
An irrational decimal fraction is created by concatenating the positive integers:  

0.123456789101112131415161718192021...  

It can be seen that the $12^{th}$ digit of the fractional part is 1.  

If $d_n$ represents the nth digit of the fractional part, find the value of the following expression.  

$d_1 × d_{10} × d_{100} × d_{1000} × d_{10000} × d_{100000} × d_{1000000}$

In [ ]:
sys.set_int_max_str_digits(maxdigits=10000000)

def champernowne_constant(n_digits):
    c = ""
    i = 1
    while len(c) <= n_digits:
        c += str(i)
        i += 1
    return c

def solve(n):
    answer = 1
    c = champernowne_constant(10 ** n)
    for i in range(n + 1):
        answer *= int(c[10 ** i - 1])
    return answer

print(f"Answer 40 = {solve(6)}")  

Answer 40 = 210


### Problem 41 - Pandigital prime
We shall say that an n-digit number is pandigital if it makes use of all the digits 1 to n exactly once. For example, 2143 is a 4-digit pandigital and is also prime.  

What is the largest n-digit pandigital prime that exists?

In [ ]:
# def is_pandigital(n, n_digits):
#     n = list(str(n))
#     if len(n) != n_digits:
#         return False
#     n = set(map(lambda x: int(x), n))
#     for i in range(1, n_digits + 1):
#         if i in n:
#             continue
#         else:
#             return False
#     return True

def is_prime(n):
    # this flag maintains status whether the n is prime or not
    prime_flag = 0
    if(n > 1):
        for i in range(2, int(math.sqrt(n)) + 1):
            if (n % i == 0):
                prime_flag = 1
                break
        if (prime_flag == 0):
            return True
        else:
            return False
    else:
        return False


def largest_pandigital_prime():
    for i in itertools.permutations(range(7, 0, -1)):
        s = functools.reduce(lambda b, a : 10 * b + a, i)
        if is_prime(s):
            return s


print(f"Answer 41 = {largest_pandigital_prime()}")

Answer 41 = 7652413


### Problem 42 - Coded triangle numbers
The nth term of the sequence of triangle numbers is given by, tn = ½n(n+1); so the first ten triangle numbers are:

1, 3, 6, 10, 15, 21, 28, 36, 45, 55, ...

By converting each letter in a word to a number corresponding to its alphabetical position and adding these values we form a word value. For example, the word value for SKY is $19 + 11 + 25 = 55 = t_{10}$. If the word value is a triangle number then we shall call the word a triangle word.

Using words.txt (right click and 'Save Link/Target As...'), a 16K text file containing nearly two-thousand common English words, how many are triangle words?

## Problems 51 to 100

### Problem 67 - Maximum path sum II
By starting at the top of the triangle below and moving to adjacent numbers on the row below, the maximum total from top to bottom is 23.  

3  
7 4  
2 4 6  
8 5 9 3  

That is, $3 + 7 + 4 + 9 = 23$.

Find the maximum total from top to bottom in triangle.txt (right click and 'Save Link/Target As...'), a 15K text file containing a triangle with one-hundred rows.  

NOTE: This is a much more difficult version of Problem 18. It is not possible to try every route to solve this problem, as there are 299 altogether! If you could check one trillion (1012) routes every second it would take over twenty billion years to check them all. There is an efficient algorithm to solve it. ;o)

In [ ]:
# drive/MyDrive/Colab\ Projects/ProjectEuler/p067_triangle.txt

with open("/content/gdrive/MyDrive/Colab_Projects/ProjectEuler/p067_triangle.txt") as f:
    triangle = f.read()

def format_triangle(tri):
    # format into lists of integers
    tri = tri.strip().split('\n')
    for i, j in enumerate(tri):
        tri[i] = [int(x) for x in j.split()]
    
    # Find the number of rows and columns
    nrows = len(tri)
    ncols = max(len(x) for x in tri)

    # extend length of all arrays to be equal
    for row in tri:
        row.extend([0] * (ncols - len(row)))

    return np.array(tri), nrows, ncols


def max_path_sum(tri, nrows, ncols, print_tri=False):
    
    for i in range(nrows-2, -1, -1):
        if print_tri:
            print(tri, "\n")
        for j in range(ncols - 1):
            if tri[i,j] != 0:
                tri[i,j] += max(tri[i+1,j], tri[i+1,j+1])
        

    return tri[0][0]


tri, nrows, ncols = format_triangle(triangle) 
print(f"Answer 67 = {max_path_sum(tri, nrows, ncols)}")

Answer 18 = 7273


## Problems 101 to 150

## Problems 151 to 200


## Problems 201 to 250

## Problems 251 to 300

## Problems 301 to 350

## Problems 351 to 400